elegy/losses/cosine_similarity_test.py

```diff
@@ -0,0 +1,112 @@
+import elegy
+
+from elegy.testing_utils import transform_and_run
+from elegy import utils
+import jax.numpy as jnp
+import jax
+from jax.lax import rsqrt
+import tensorflow.keras as tfk
+
+# import debugpy
+
+# print("Waiting for debugger...")
+# debugpy.listen(5679)
+# debugpy.wait_for_client()
+
+
+@transform_and_run
+def test_basic():
+
+    y_true = jnp.array([[0.0, 1.0], [1.0, 1.0]])
+    y_pred = jnp.array([[1.0, 0.0], [1.0, 1.0]])
+
+    # Using 'auto'/'sum_over_batch_size' reduction type.
+    cosine_loss = elegy.losses.CosineSimilarity(axis=1)
+    assert cosine_loss(y_true, y_pred) == -0.49999997
+
+    # Calling with 'sample_weight'.
+    assert (
+        cosine_loss(y_true, y_pred, sample_weight=jnp.array([0.8, 0.2])) == -0.099999994
+    )
+
+    # Using 'sum' reduction type.
+    cosine_loss = elegy.losses.CosineSimilarity(
+        axis=1, reduction=elegy.losses.Reduction.SUM
+    )
+    assert cosine_loss(y_true, y_pred) == -0.99999994
+
+    # Using 'none' reduction type.
+    cosine_loss = elegy.losses.CosineSimilarity(
+        axis=1, reduction=elegy.losses.Reduction.NONE
+    )
+
+    assert jnp.equal(cosine_loss(y_true, y_pred), jnp.array([-0.0, -0.99999994])).all()
+
+
+@transform_and_run
+def test_function():
+
+    rng = jax.random.PRNGKey(42)
+
+    y_true = jax.random.randint(rng, shape=(2, 3), minval=0, maxval=2)
+    y_pred = jax.random.uniform(rng, shape=(2, 3))
+
+    def _l2_normalize(x, axis=None, epsilon=utils.EPSILON):
+        square_sum = jnp.sum(jnp.square(x), axis=axis, keepdims=True)
+        x_inv_norm = rsqrt(jnp.maximum(square_sum, epsilon))
+        return jnp.multiply(x, x_inv_norm)
+
+    loss = elegy.losses.cosine_similarity(y_true, y_pred, axis=1)
+
+    assert loss.shape == (2,)
+
+    y_true = _l2_normalize(y_true, axis=1)
+    y_pred = _l2_normalize(y_pred, axis=1)
+    assert jnp.array_equal(loss, -jnp.sum(y_true * y_pred, axis=1))
+
+
+@transform_and_run
+def test_compatibility():
+    # Input:  true (y_true) and predicted (y_pred) tensors
+    y_true = jnp.array([[0.0, 1.0], [0.0, 0.0]])
```

Comment:
Can we use random arrays of higher dimensions here so the test is more general? We can use `np.random.uniform` to create them.

Reply:
I guess there is a bug in TF implementation in ```CosineSimilarity```. I've raise an issue TF here, https://github.com/tensorflow/tensorflow/issues/44343

Reply:
I've updated the PR kindly have a look.